.github/scripts/manage-vm.py

```diff
@@ -98,36 +98,70 @@ def generate_cloud_init_script(
             f",GITHUB_REPOSITORY_{os.environ['GITHUB_REPOSITORY'].replace('/', '_')}"
         )
 
-    for item in ["GITHUB_SHA", "GITHUB_REF", "GITHUB_RUN_ID", "GITHUB_RUN_NUMBER"]:
+    for item in ["GITHUB_SHA", "GITHUB_REF", "GITHUB_RUN_ID", "GITHUB_RUN_ATTEMPT"]:
         if os.environ.get(item):
             label += f",{item}_{os.environ[item]}"
 
-    script = [
-        "#!/bin/bash",
-        "set -x",
-        "mkdir -p /actions-runner && cd /actions-runner",
-        'case $(uname -m) in aarch64) ARCH="arm64" ;; amd64|x86_64) ARCH="x64";; esac',
-        f"export FILENAME=actions-runner-linux-${{ARCH}}-{version}.tar.gz",
-        # https://github.com/actions/runner/releases/download/v2.314.1/actions-runner-linux-x64-2.314.1.tar.gz
-        f'[ -f "$FILENAME" ] || curl -O -L "https://github.com/actions/runner/releases/download/v{version}/$FILENAME"',
-        'tar xzf "./$FILENAME"',
-        "export RUNNER_ALLOW_RUNASROOT=1",
-        f"./config.sh  --labels {label} --url https://github.com/{owner}/{repo} --token {token}",
-        "./run.sh",
-    ]
+    script = f"""
+set -x
+[ -d /actions-runner ] && {{
+    echo "Runner already installed"
+    cd /actions-runner
+}} || {{
+    mkdir -p /actions-runner && cd /actions-runner
+    case $(uname -m) in
+        aarch64) ARCH="arm64" ;;
+        amd64|x86_64) ARCH="x64";;
+    esac
+    export FILENAME=runner.tar.gz
+    # https://github.com/actions/runner/releases/download/v2.314.1/actions-runner-linux-x64-2.314.1.tar.gz
+    exit_code=1
+    i=0
+    url="https://github.com/actions/runner/releases/download/v{version}/actions-runner-linux-${{ARCH}}-{version}.tar.gz"
+    until [ $exit_code -eq 0 ] || [ $i -gt 3 ]; do
+        [ -f "$FILENAME" ] || curl --connect-timeout 5 -O -L "$url" -o "$FILENAME"
+        exit_code=$?
+        i=$((i+1))
+        [ $exit_code -eq 0 ] || rm -f "$FILENAME"
+        echo "$((date)) [$i] curl exited (or timed-out) with code $exit_code"
+    done
+    tar xzf "./$FILENAME" || exit 0
+}}
+export RUNNER_ALLOW_RUNASROOT=1
+
+# trying to catch registration error
+exit_code=1
+i=0
+until [ $exit_code -eq 0 ] || [ $i -gt 3 ]; do
+    echo ./config.sh --labels {label} --url https://github.com/{owner}/{repo} --token XXX --unattended
+    set +x
+    timeout 60 ./config.sh --labels {label} --url https://github.com/{owner}/{repo} --token {token} --unattended
+    set -x
+    exit_code=$?
+    i=$((i+1))
+    echo "$((date)) [$i] config.sh exited (or timed-out) with code $exit_code"
+    [ $exit_code -eq 0 ] || find /actions-runner -name *.log -print -exec cat {{}} \; # noqa: W605
+done
+# exit code 0 to skip the error and to boot vm correctly
+./run.sh || exit 0
+"""
 
     cloud_init = {}
-    cloud_init["ssh_pwauth"] = False
+    # cloud_init["ssh_pwauth"] = False
```

Comment:
```
**(ssh_keys and {"ssh_keys": ssh_keys} or {})
```
Also do not need to assign ["run_cmd"]. You can make it a single dict literal.

Reply:
I am not really familiar with this syntax and i try not to use esoteric syntax when I can, because it complicates things for people who are not familiar with it either (ie c++ developers who may not be familiar with python at all). And it is not some cpu/memory intensive application to go full swing for those optimisations with memory allocation where readability is more important.